File: java/math/EratosthenesSieve.java
package math;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class EratosthenesSieve {
    public int[] sieve(int num) {
        boolean[] isPrime = new boolean[num];
        Arrays.fill(isPrime, true);
        List<Integer> result = new ArrayList<>();
        for (int i = 2; i <= num; ++i) {
            if (isPrime[i - 1]) {
                result.add(i);
            }

            for (int value : result) {
                if (value > num / i) {
                    break;
                }
                isPrime[i * value - 1] = false;
            }
        }
        return result.stream().mapToInt(Integer::valueOf).toArray();
    }
}

File: cpp/math/EratosthenesSieve.cpp
#include <iostream>
#include <vector>
using namespace std;

class EratosthenesSieve {
public:
    vector<int> sieve(int num) {
        vector<bool> isPrime(num, true);
        vector<int> result(0);
        for (int i = 2; i <= num; ++i) {
            if (isPrime[i - 1]) {
                result.emplace_back(i);
            }

            for (int value : result) {
                if (value > num / i) {
                    break;
                }
                isPrime[i * value - 1] = false;
            }
        }
        return result;
    }
};
